resources/testharness.js

```diff
@@ -4296,81 +4287,84 @@
             if (!asserts) {
-                return "No asserts ran";
+                asserts_output.querySelector("summary").textContent = "No asserts ran";
+                return asserts_output;
             }
-            rv = "<table>";
-            rv += asserts.map(assert => {
-                var output_fn = "<strong>" + escape_html(assert.assert_name) + "</strong>(";
-                var prefix_len = output_fn.length;
-                var output_args = assert.args;
-                var output_len = output_args.reduce((prev, current) => prev+current, prefix_len);
-                if (output_len[output_len.length - 1] > 50) {
-                    output_args = output_args.map((x, i) =>
-                    (i > 0 ? "  ".repeat(prefix_len) : "" )+ x + (i < output_args.length - 1 ? ",\n" : ""));
-                } else {
-                    output_args = output_args.map((x, i) => x + (i < output_args.length - 1 ? ", " : ""));
-                }
-                output_fn += escape_html(output_args.join(""));
-                output_fn += ')';
-                var output_location;
+
+            const table = asserts_output.querySelector("table");
+            const row = render(
+                    ["tr", {},
+                        ["td", {}, ""],
+                        ["td", {}, ["pre", {}, ["strong", {}], ""]] ]);
+            const row_status = row.querySelectorAll("td")[0];
+            const row_name = row.querySelector("strong");
+            const row_other_info = row_name.nextSibling;
+            for (const assert of asserts) {
+                var status_class_name = status_class(Test.prototype.status_formats[assert.status]);
+                row.className = "overall-" + status_class_name;
+                row_status.className = status_class_name;
+                row_status.textContent = Test.prototype.status_formats[assert.status];
+                row_name.textContent = assert.assert_name;
+                var output_fn = "(" + assert.args.join(", ") + ")";
                 if (assert.stack) {
-                    output_location = assert.stack.split("\n", 1)[0].replace(/@?\w+:\/\/[^ "\/]+(?::\d+)?/g, " ");
+                    output_fn += "\n";
+                    output_fn += assert.stack.split("\n", 1)[0].replace(/@?\w+:\/\/[^ "\/]+(?::\d+)?/g, " ");
                 }
-                return "<tr class='overall-" +
-                    status_class(Test.prototype.status_formats[assert.status]) + "'>" +
-                    "<td class='" +
-                    status_class(Test.prototype.status_formats[assert.status]) + "'>" +
-                    Test.prototype.status_formats[assert.status] + "</td>" +
-                    "<td><pre>" +
-                    output_fn +
-                    (output_location ? "\n" + escape_html(output_location) : "") +
-                    "</pre></td></tr>";
+                row_other_info.textContent = output_fn;
+
+                table.appendChild(row.cloneNode(true));
```

Comment:
Is there a reason behind using `cloneNode` here? If so, please document.

Reply:
Hm, I don't see it. You're moving some element creation into the browser implementation (I'd guess string copies under this would be refcounted), but I don't think you're actually saving any `textContent` sets. The actual perf impact is anyone's guess, of course. If you've got time to compare, I think using `render` directly would be more readable, at least.

Reply:
Updated the PR to use `render()`. It certainly is more readable.